language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -55,48 +124,112 @@ type VarEnv = UniqueMap TermUnique Closure
+    spendWMemory _ _ = pure ()
+    spendWCPU _ _ = pure ()
+
+class SpendExUnitsState s where
+    spendSMemory :: forall w. Monoid w => (WithMemory Term) -> ExMemory -> CekMBase w s ()
+    spendSCPU :: forall w. Monoid w => (WithMemory Term) -> ExCPU -> CekMBase w s ()
+
+instance SpendExUnitsState () where
+    spendSCPU _ _ = pure ()
+    spendSMemory _ _ = pure ()
+
+spendSBudget :: Ord a => Num a => Monoid w => Lens' ExBudget a -> WithMemory Term -> a -> CekMBase w ExBudget ()
+spendSBudget l term ex = do
+    newEx <- fmap (\x -> x - ex) $ gets (view l)
+    if newEx < 0 then
+        throwError $ MachineException OutOfExError (void term)
+    else
+        assign l newEx
+
+instance SpendExUnitsState ExBudget where
+    spendSCPU = spendSBudget exBudgetCPU
+    spendSMemory = spendSBudget exBudgetMemory
 
 data Frame
-    = FrameApplyFun VarEnv (Plain Value)               -- ^ @[V _]@
-    | FrameApplyArg VarEnv (Plain Term)                -- ^ @[_ N]@
-    | FrameTyInstArg (Type TyName ())                  -- ^ @{_ A}@
+    = FrameApplyFun VarEnv (WithMemory Value)               -- ^ @[V _]@
+    | FrameApplyArg VarEnv (WithMemory Term)                -- ^ @[_ N]@
+    | FrameTyInstArg (Type TyName ExMemory)                  -- ^ @{_ A}@
     | FrameUnwrap                                      -- ^ @(unwrap _)@
-    | FrameIWrap () (Type TyName ()) (Type TyName ())  -- ^ @(iwrap A B _)@
+    | FrameIWrap ExMemory (Type TyName ExMemory) (Type TyName ExMemory)  -- ^ @(iwrap A B _)@
 
 type Context = [Frame]
 
 makeLenses ''CekEnv
 
-runCekM :: CekEnv -> CekM a -> Either CekMachineException a
-runCekM = flip runReaderT
+runCekM :: SpendExUnits w s => CekEnv -> s -> CekM w s a -> (Either CekMachineException a, s, w)
+runCekM r s cekm = runRWS (runExceptT cekm) r s
 
 -- | Get the current 'VarEnv'.
-getVarEnv :: CekM VarEnv
+getVarEnv :: CekM w s VarEnv
 getVarEnv = asks _cekEnvVarEnv
 
 -- | Set a new 'VarEnv' and proceed.
-withVarEnv :: VarEnv -> CekM a -> CekM a
-withVarEnv = local . set cekEnvVarEnv
+withVarEnv :: VarEnv -> CekM w s a -> CekM w s a
+withVarEnv venv = mapExceptT (local (set cekEnvVarEnv venv))
 
 -- | Extend an environment with a variable name, the value the variable stands for
 -- and the environment the value is defined in.
-extendVarEnv :: Name () -> Plain Value -> VarEnv -> VarEnv -> VarEnv
+extendVarEnv :: Name ExMemory -> WithMemory Value -> VarEnv -> VarEnv -> VarEnv
 extendVarEnv argName arg argVarEnv = insertByName argName $ Closure argVarEnv arg
 
 -- | Look up a variable name in the environment.
-lookupVarName :: Name () -> CekM Closure
+lookupVarName :: Name ExMemory -> CekM w s Closure
 lookupVarName varName = do
     varEnv <- getVarEnv
     case lookupName varName varEnv of
-        Nothing   -> throwError $ MachineException OpenTermEvaluatedMachineError (Var () varName)
+        Nothing   -> throwError $ MachineException OpenTermEvaluatedMachineError (Var () (void varName))
```

Comment:
This line made me realise that `Closure` is still using un-annotated values. Why not use annotated values? Plus, I think you need that to do the strategy we discussed where you look at the annotated memory values for things in the environment.